### ISSUE TITLE
Improving the commandline

### ISSUE THREAD

**Issue by cmcaine** [cmcaine/tridactyl#27]
The commandline needs history, autocomplete, readline-style bindings (C-u, C-w, etc), and more

Koushien's working on completion, but the general architecture and the rest of the work is up for grabs.

Here's a project that has an at-first-glance sensible architecture (the middleware stuff) and may have code we can nick: https://www.npmjs.com/package/shell

Blocks #24.

**Comment by cmcaine** [comment]
History now working, mostly thanks to Olie. I've made it globally stored in 21c06d9520910a55fb30d1a19197949211d92ae8

Whole commandline is a bit janky still, though

**Comment by bovine3dom** [comment]
History search on up/down added and clamping fixed in 11155e8.

**Comment by KrazyCavin** [comment]
It would be awesome if it could support "tab" to autocomplete command, like:
```
he<tab> -> help
searchs<tab> -> searchsetkeyword
```

**Comment by bovine3dom** [comment]
We were thinking of having an excmd completion category, like we do for tabs and open. We'd have to think about how to fill in the command line though. You can currently use `Ctrl+F` to complete from history, if that helps.

**Comment by cmcaine** [comment]
@behrmann wants to be able to scroll the bufferlist

> I usually have more than 19 tabs open at a time. When pressing b I get a list of buffers 1 through 19 which I can tab through, unfortunately the tabbing wraps at 19 and goes back to tab 1. The list of course be moved to the actual tabs one wishes to see by typing keywords, but a nicer solution would be to scroll the list further when tabbing beyond the bottom entry and only wrap at the last tab.

**Comment by Koushien** [comment]
Re: The post immediately above this one
You probably have your own notes on this, but for everyone interested Vimperator's default behavior updates the bufferlist view when there exists an item we can cycle towards before wrapping around that is 3 cycles away.
That is, suppose I have more than 20 buffers open and can only display at most 20 buffers in the list. If I press tab to cycle forwards beginning at the first buffer, it only shows buffers 1 through 20 until I reach buffer 18 in the list, where it would then show buffers 2-21. If I begin cycling backwards from there, buffers 1 through 20 will only display when I reach buffer 4.

I'm not advocating Vimperator's approach is the best, just thought I'd share that in case someone else wanted to take a stab at this.

**Comment by madduck** [comment]
FWIW, in Vim, this is controlled by the `scrolloff` setting, and it's indeed very nice.

**Comment by bovine3dom** [comment]
@alem0lars nothing unreleased. It's blocked by us wanting to rewrite the command line to be a bit less messy.

**Comment by alem0lars** [comment]
@bovine3dom Ok thanks. Is that feature already planned or started?

**Comment by cmcaine** [comment]
Planned, not really started. If you're interested in collaborating I can write up/find our current design thoughts for you.

**Comment by bovine3dom** [comment]
@alem0lars which feature? If you mean the rewrite, it hasn't been started or really planned yet, other than that we think we should probably use Inferno. @cmcaine will know more :)

**Comment by cmcaine** [comment]
Heh, conflicting responses. It's vaguely planned at least :)

**Comment by bovine3dom** [comment]
Closing a bunch of PRs which look unlikely to be finished but should prove as prior art for anyone brave enough to tackle the commandline rewrite:

#611 

#932 

#1270

**Comment by arcnmx** [comment]
Would this also encompass an improvement to excommand argument parsing, or would that be a separate issue? (as in, a system that also enables things like quotes for providing arguments with spaces etc, maybe getopt style...)

**Comment by glacambre** [comment]
@arcnmx The parser has its own issue: https://github.com/tridactyl/tridactyl/issues/789
